lab/lab-1.md

```diff
@@ -0,0 +1,209 @@
+[centos@node-0 example-app-frontend]$ git checkout v2.0.0
+Previous HEAD position was fc0df8a fix warning
+HEAD is now at dc65f9b Change background color to green
+[centos@node-0 example-app-frontend]$
+```
+
+Now just build a new version of the frontend `v2.0.0` with the provided Dockerfile. And push the image to DockerHub. Use again your account.
+
+### Deploy frontend v2.0.0
+
+So now create again a YAML file `deployment_v2.yaml` for the new version and deploy it to the cluster. Don't forget to adapt the labels and the name. We don't want to perform an update on the exsiting deployment.
+
+```sh
+[centos@node-0 example-app-frontend]$ kubectl apply -f deployment_v2.yaml
+```
+
+You should see now 3 deployments(`backend-v1`, `frontend-v1`, `frontend-v2`):
+
+```sh
+[centos@node-0 example-app-frontend]$ kubectl get deployments --namespace demoapp
+[centos@node-0 example-app-frontend]$
+```
+
+TODO insert sample output (!)
+
+### Release new version
+
+Access the frontend again in your browser. Has the background color changed to green?
+
+No, we first have to switch the service to the new version. Therfore update the `service.yaml` of the frontend and target only the pods with version `v2.0.0`.
+Apply the manifest:
+
+```sh
+[centos@node-0 example-app-frontend]$ kubectl apply -f service.yaml
+```
+
+Now reload the page in your browser. The background color should have changed from blue to green and our customers are now happy.
+
+## Canary releases
+
+Our development team has built a new feature, but we just want to test the new feature with half of our users. Therefore we want to run both version(the stable v1.0.0 and the new version) in parallel.
+
+### Build new backend - v2.0.0
+
+First we have to build the new version and push it to our docker registry at DockerHub.
+
+```sh
+[centos@node-0 ~]$ cd
+[centos@node-0 ~]$ cd example-app-backend
+[centos@node-0 example-app-backend]$ git checkout v2.0.0
+Previous HEAD position was a16d5f4 Update go.yml
+HEAD is now at 0ca7ba1 Return version 2.0.0
+[centos@node-0 example-app-backend]$
+```
+
+Now just build a new version of the backend `v2.0.0` with the provided Dockerfile. And push the image to DockerHub. Use again your account.
+
+### Deploy new backend to the cluster
+
+So now create again a YAML file `deployment_v2.yaml` for the new version and deploy it to the cluster. Don't forget to adapt the labels and the name. We don't want to perform an update on the exsiting deployment.
+
+```sh
+[centos@node-0 example-app-backend]$ kubectl apply -f deployment_v2.yaml
+```
+
+You should see now 4 deployments(`backend-v1`, `backend-v1`, `frontend-v1`, `frontend-v2`):
+
+```sh
+[centos@node-0 example-app-backend]$ kubectl get deployments --namespace demoapp
+[centos@node-0 example-app-backend]$
+```
+
+TODO insert sample output (!)
+
+### Partially release new version
+
+Access the frontend in your browser and reload it a few times. Is it always showing backend version `1.0.0` or sometimes `2.0.0`?
```

Comment:
is it `v1.0.0` or `1.0.0`?

Reply:
it is 1.0.0, v1.0.0 is the version I am using for image tagging